### ISSUE TITLE
Parallelize Requesting of Availability Chunks in Distribution

### ISSUE THREAD

**Issue by rphmeier** [paritytech/polkadot#2512]
re: https://github.com/paritytech/polkadot/pull/2423#discussion_r580634082

If validators are actually online and responding quickly whether or not they have the chunk, then parallelization is not important. However, if there are validators who are offline or not responding to chunk requests, the serial requesting can cause delays in parachain availability and thus in throughput.

Some optimistic parallelization of requests would waste bandwidth but create better throughput for parachains.

**Comment by eskimor** [comment]
I think we should really not be doing that. availability-distribution is not time critical and network bandwidth is a scarce resource. Instead I would like to optimize the serial requesting:

1. Have a queue management similar to statement distribution: Fail quickly if we already know that we are likely not be able to respond in a timely manner.
2. Have relative short timeouts.
3. Prioritize validators in request handling

Disconnected peers are also no longer a problem as we are now connected to all validators by default - by disabling connect on disconnect, we would not even try an offline peer, therefore also failing quickly.

1. is particular nice, because it does automatic and dynamic load balancing. Therefore evening out any randomness bias.

If there are no objections, I would change the name and goal of this ticket to something like  "optimizing serial transfer".

**Comment by burdges** [comment]
Yes.  After backing but before availability, backers only send chunk k to validator k during availability distribution for all k, even other validators should not request chunks besides their own.  

After availability, reconstructing nodes could either request chunk k from validator k for many k, or else request the entire block from anyone who signed they possess it, and then fall back to chunks when this fails.  We clearly prioritize validator at this stage, maybe prioritize assigned validators.  

It's fine if reconstruction has several requests in flight simultaneously of course, especially early during reconstruction.  If however we eventually reward correct availability behavior via the tit-for-tat strategy described in https://github.com/paritytech/polkadot/pull/2089#issuecomment-740503318 and https://github.com/w3f/research-security-issues/issues/30#issuecomment-768593265 then validators wind up losing some rewards for requesting more than the reconstruction threshold:  Alice obtains all but one chunk, so she asks Bob and Carol for chunks, who both respond.  Alice sees Bob's chunk first, so she credits Bob.  Carol thinks Alice stiffed her, so Carol skips requesting from Alice sometime soon, slightly hurting Alice's rewards.
